### PR TITLE
refactor(fetcher): clean up fetcher

### PR DESCRIPTION
**Pull request by jaycxz**

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
-            "name": "Python: Module",
-            "type": "python",
-            "request": "launch",
-            "module": "src.ohlc_fetcher_initializer",
-            "justMyCode": true
-        }
-    ]
-}
\ No newline at end of file
+  // Use IntelliSense to learn about possible attributes.
```
Comment (MingzXIE) [inline .vscode/launch.json]: add this folder to gitignore

File: src/utils/ohlc_fetcher.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+import datetime as dt
+import json
+import time
+
+import pandas as pd
+import requests
+from sqlalchemy import create_engine, desc, select
+
+engine = create_engine(
+    "postgresql+psycopg2://Jayc:admin@127.0.0.1:5432/bigaluto_stockdb"
```
Comment (MingzXIE) [inline src/utils/ohlc_fetcher.py]: encapsulate the connection string into environment environments with default value

File: src/sql/schema.py [hunk 3]
```diff
@@ -1,28 +1,33 @@
...
 class Data(Base):
     __tablename__ = "data"
 
     date: Mapped[DATE] = mapped_column(DATE, primary_key=True)
     ticker_id = mapped_column(Integer(), ForeignKey("ticker.id"), primary_key=True)
-    timeframe_id = mapped_column(Integer(), ForeignKey("timeframe.id"), primary_key=True)
+    timeframe_id = mapped_column(
+        Integer(), ForeignKey("timeframe.id"), primary_key=True
+    )
     open: Mapped[Float] = mapped_column(Float)
```
Comment (MingzXIE) [inline src/sql/schema.py:57]: open_price?

Comment (MingzXIE) [reply src/sql/schema.py:57]: or price_open?

File: src/sql/schema.py [hunk 4]
```diff
@@ -1,28 +1,33 @@
 from typing import List
 
-from sqlalchemy import DATE, BigInteger, Column, Float, ForeignKey, Integer, String
+from sqlalchemy import DATE, BigInteger, Float, ForeignKey, Integer, String
 from sqlalchemy.orm import Mapped, declarative_base, mapped_column, relationship
 
 Base = declarative_base()
 
+
```
Comment (MingzXIE) [inline src/sql/schema.py:8]: please add Doc strings for the below tables

File: src/utils/ohlc_fetcher.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    timeframe_win = "D"
+
+    for date, ticker_list in ticker_collection.items():
+        start_date = dt.datetime.strptime(date, "%Y-%m-%d").date()
+        for ticker_id, _ in ticker_list:
+            try:
+                unixts_1 = int(calendar.timegm(start_date.timetuple()))
+                unixts_2 = int(calendar.timegm(dt.datetime.utcnow().timetuple()))
+
+                query = f"https://query2.finance.yahoo.com/v8/finance/chart/NVR?formatted=true&crumb=0VkTr5j7Fso&lang=en-GB&region=US&includeAdjustedClose=true&interval=1d&period1={unixts_1}&period2={unixts_2}&events=capitalGain%7Cdiv%7Csplit&useYfid=true&corsDomain=uk.finance.yahoo.com"
```
Comment (MingzXIE) [inline src/utils/ohlc_fetcher.py]: create a constants.py file and move the url into it.

File: src/utils/ohlc_fetcher.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+engine = create_engine(
+    "postgresql+psycopg2://Jayc:admin@127.0.0.1:5432/bigaluto_stockdb"
+)
+
+from sqlalchemy.orm import Session
+
+from ..sql.schema import Data, Ticker, Timeframe
+
+
+def execute_fetcher(
```
Comment (MingzXIE) [inline src/utils/ohlc_fetcher.py]: please add doc str

File: src/utils/ohlc_fetcher.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    ticker_collection: dict[str, list[tuple[int, str]]],
+):
+    headers = {
+        "Accept": "*/*",
+        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36",
+    }
+
+    all_errors = []
+    timeframe_win = "D"
```
Comment (MingzXIE) [inline src/utils/ohlc_fetcher.py]: is this the default timeframe window?

File: src/utils/ohlc_fetcher.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+                        "high": ohlc_content["high"],
+                        "low": ohlc_content["low"],
+                        "close": ohlc_content["close"],
+                        "adj_close": adj_close["adjclose"],
+                        "volume": ohlc_content["volume"],
+                    }
+                )
+                df["date"] = pd.to_datetime(df["date"], unit="s").dt.date
+
+                with Session(engine) as session:
```
Comment (MingzXIE) [inline src/utils/ohlc_fetcher.py]: make the DB operations more generic and put into util file, so we can simply use db_util.create_enitites() / db_util.delete_entities() etc. to interact with DB in the functions

File: src/utils/ohlc_fetcher.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+    for date, ticker_list in ticker_collection.items():
+        start_date = dt.datetime.strptime(date, "%Y-%m-%d").date()
+        for ticker_id, _ in ticker_list:
+            try:
+                unixts_1 = int(calendar.timegm(start_date.timetuple()))
+                unixts_2 = int(calendar.timegm(dt.datetime.utcnow().timetuple()))
+
+                query = f"https://query2.finance.yahoo.com/v8/finance/chart/NVR?formatted=true&crumb=0VkTr5j7Fso&lang=en-GB&region=US&includeAdjustedClose=true&interval=1d&period1={unixts_1}&period2={unixts_2}&events=capitalGain%7Cdiv%7Csplit&useYfid=true&corsDomain=uk.finance.yahoo.com"
+                r = requests.get(query, headers=headers)
```
Comment (MingzXIE) [inline src/utils/ohlc_fetcher.py]: use a try - except caluse to catch the potential api call errors

File: src/utils/ohlc_fetcher.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+                content = r.json()
+
+                content_data = content["chart"]["result"][0]
```
Comment (MingzXIE) [inline src/utils/ohlc_fetcher.py]: please handle the case that 'chart' / 'result' field do not exist

File: src/utils/ohlc_fetcher.py (hunk 9 above)
Comment (MingzXIE) [inline src/utils/ohlc_fetcher.py]: can use multi-thread to make async calls

**Review by MingzXIE — CHANGES_REQUESTED** [review]

**Comment by MingzXIE** [comment]
Good start, just few comments/queries

File: src/sql/schema.py (hunk 3 above)
Comment (jaycxz) [reply src/sql/schema.py:57]: price at the open of market

File: src/utils/ohlc_fetcher.py (hunk 7 above)
Comment (jaycxz) [reply src/utils/ohlc_fetcher.py]: yeah

**Comment by jaycxz** [comment]
@MingzXIE , updated commit. The only thing i have not updated is to make it async. Will see to it tmr or something

**Pull request merged**